### PR TITLE
feat(profiling): allow to export memory heap to Python

### PR DESCRIPTION
**Pull request by jd**
Adds a _memalloc.heap() method that returns a list of traceback using memory
right now with their allocated size.

### CODE REVIEW

**Comment by jd** [comment]
Re-reading the code made me think about something about the freezer, switching to draft for now.

**Comment by jd** [comment]
Ok, there was a potential bug in the way things were done on thaw.

The code would handle the buffered `frees` from the freezer on the existing allocation of the heap tracker, and then merge the allocations with the freezer allocation.

The problem is that in the (unlikely) case a `free` from the freezer would match an `alloc` from the freezer, it would not have been removed at all since the `heap_tracker_thaw` was only handling the freezer frees against the heap tracker allocs — it was missing the freezer allocs!

I've fixed that by inverting the two operations: we first merge all the allocations together, and then we handle the free for the whole reunited allocation array.

File: ddtrace/profiling/collector/_memalloc_heap.c [hunk 1]
```diff
@@ -56,12 +65,72 @@ memalloc_heap_untrack(void* ptr)
...
+memalloc_heap_tracker_thaw()
+{
+    heap_tracker_thaw(&global_heap_tracker);
+}
+
+void
+memalloc_heap_untrack(void* ptr)
+{
+    if (global_heap_tracker.frozen)
+        /* FIXME check for max capacity of frees */
```
Comment (Kyle-Verhoog) [inline ddtrace/profiling/collector/_memalloc_heap.c:128]: Worth doing still to be safe? It is unlikely that there will be very many frees while frozen so probably not a big concern, but then again, shouldn't be too hard to check either 🤷‍♂️

File: ddtrace/profiling/collector/_memalloc_heap.c [hunk 2]
```diff
@@ -56,12 +65,72 @@ memalloc_heap_untrack(void* ptr)
...
+heap_tracker_thaw(heap_tracker_t* heap_tracker)
+{
+    /* Add the frozen allocs at the end */
+    traceback_array_splice(&heap_tracker->allocs,
+                           heap_tracker->allocs.count,
+                           0,
+                           heap_tracker->freezer.allocs.tab,
+                           heap_tracker->freezer.allocs.count);
+
+    /* Handle the frees: we need to handle the frees after we merge the allocs
```
Comment (Kyle-Verhoog) [inline ddtrace/profiling/collector/_memalloc_heap.c:84]: :+1:

**Review by Kyle-Verhoog — APPROVED** [review]
just one comment but it might be reasonable to pass on it, I'll leave the decision up to you 😆

**Comment by Kyle-Verhoog** [comment]
oh right forgot about automerge.. so I guess I made the decision.... but then CI is probably going to flake out.... so you get to decide.... maybe 😆

**Comment by jd** [comment]
I'll handle the FIXME, better be safe than sorry.

**Review by Kyle-Verhoog — APPROVED** [review]

**Comment by jd** [comment]
@Kyle-Verhoog I need to re-approve on this

**Review by Kyle-Verhoog — APPROVED** [review]

**Pull request merged**